include/aws/common/array_list.h

```diff
@@ -0,0 +1,265 @@
+#ifndef AWS_COMMON_ARRAY_LIST_H
+#define AWS_COMMON_ARRAY_LIST_H
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <aws/common/error.h>
+#include <stdint.h>
+#include <string.h>
+
+/* To declare an instance of a typed array use this macro (it's largely for header files). Then use the definition macro
+ * for your .c file */
+#define aws_typed_array_list_declaration(name, T)                                                                      \
+struct aws_##name##_array_list {                                                                                       \
+    struct aws_allocator *alloc;                                                                                       \
+    size_t current_size;                                                                                               \
+    size_t length;                                                                                                     \
+                                                                                                                       \
+    T *data;                                                                                                           \
+};
+
+/* To define your typed array list, use this macro (most likely in your .c file) */
```

Comment:
Both 'declaration' and 'definition' will cause compile errors if inserted twice in the same translation unit. Also, the functions are `static inline` so need to be included in each translation unit that use them anyway. Why not combine them?

Reply:
DECLARATION can go in your header file without bloating it
DEFINITION can be used in your .c file

They'll only be a compile error if you declare same name and type multiple times, and in that case I'm not sure I care.

Reply:
What if I want to expose an array list as something I pass as an out/in parameter to a function in a header file? Who is responsible for declaring it?

If I do declare it in my header, only my one .c file can use that type, because I can only 'define' it once in my .c file, right? And since the definitions are all static, I can't even redeclare the functions manually to use them elsewhere.

Normally with C, things can be "declared" more than once, but only "defined" once. Let's not make our own definitions of those terms, and let's make sure we handle the very basic use case of "I want this structure to be passed between two .c files and manipulated by both".

Reply:
eh, if we get rid of the template this whole thing can just go away anyways....